cmd/frontend/graphqlbackend/cody_keyword_context.go

```diff
@@ -0,0 +1,265 @@
+
+func (r resultDocuments) Len() int           { return len(r) }
+func (r resultDocuments) Swap(i, j int)      { r[i], r[j] = r[j], r[i] }
+func (r resultDocuments) Less(i, j int) bool { return r[i].score < r[j].score }
+
+type documentStats struct {
+	TotalFiles     int
+	TermTotalFiles map[string]int
+	FileTermsCount map[string]map[string]int
+	FileBytesCount map[string]int
+}
+
+func findStats(terms *queryTerms, documents resultDocuments) *documentStats {
+	stats := documentStats{}
+
+	stats.TotalFiles = len(documents)
+	stats.TermTotalFiles = map[string]int{}
+	stats.FileTermsCount = map[string]map[string]int{}
+	stats.FileBytesCount = map[string]int{}
+
+	for _, document := range documents {
+		for _, term := range *terms {
+			r := regexp.MustCompile(fmt.Sprintf("(?i)(?:%s)", strings.Join(term.Terms(), "|")))
+			matchesCount := len(r.FindAllString(document.content, -1))
+
+			if matchesCount > 0 {
+				stats.TermTotalFiles[term.stem] += 1
+			}
+
+			fileTerms := stats.FileTermsCount[document.path]
+			if fileTerms == nil {
+				fileTerms = map[string]int{}
+			}
+
+			fileTerms[term.stem] = matchesCount
+			stats.FileTermsCount[document.path] = fileTerms
+		}
+
+		stats.FileBytesCount[document.path] = len(document.content)
+	}
+
+	return &stats
+}
+
+type queryTerm struct {
+	base        string
+	stem        string
+	escaped     string
+	escapedStem string
+}
+
+func (t *queryTerm) Terms() []string {
+	ts := []string{}
+
+	if t.base != "" {
+		ts = append(ts, t.base)
+
+		if t.base != t.stem && t.stem != "" {
+			ts = append(ts, t.stem)
+		}
+
+		if t.base != t.escaped && t.escaped != "" {
+			ts = append(ts, t.escaped)
+
+			if t.escaped != t.escapedStem && t.escapedStem != "" {
+				ts = append(ts, t.escapedStem)
+			}
+		}
+	}
+
+	return ts
+}
+
+type queryTerms []*queryTerm
+
+func (ts *queryTerms) All() []string {
+	terms := []string{}
+
+	for _, t := range *ts {
+		terms = append(terms, t.Terms()...)
+	}
+
+	return terms
+}
+
+func (ts *queryTerms) Stems() []string {
+	stems := []string{}
+
+	for _, t := range *ts {
+		stem := t.stem
+		if stem != "" {
+			stems = append(stems, t.stem)
+		}
+	}
+
+	return stems
+}
+
+func termFromStr(part string) *queryTerm {
+	base := strings.ToLower(strings.TrimSpace(stopwords.CleanString(part, "en", true)))
+	stem := stemmer.StemString(base)
+	escaped := strings.ToLower(strings.TrimSpace(regexp.MustCompile(`[$()*+./?[\\\]^{|}-]`).ReplaceAllString(part, `\$0`)))
+	escapedStem := stemmer.StemString(escaped)
+
+	return &queryTerm{
+		base:        base,
+		stem:        stem,
+		escaped:     escaped,
+		escapedStem: escapedStem,
+	}
+}
+
+func termsFromPath(path string) *queryTerms {
+	terms := queryTerms{}
+	for _, part := range regexp.MustCompile(`[/.]`).Split(path, -1) {
+		terms = append(terms, termFromStr(part))
+	}
+	return &terms
+}
+
+func idf(termTotalFiles map[string]int, totalFiles int) map[string]float64 {
+	logTotal := math.Log(float64(totalFiles))
+	idf := map[string]float64{}
```

Comment:
nit: we know the size of this ahead of time, so we can precalculate it

```suggestion
	idf := make(map[string]float64, len(termTotalFiles))
```